src/adaptors/mod.rs

```diff
@@ -243,6 +243,23 @@ impl<I> Iterator for PutBack<I>
         size_hint::add_scalar(self.iter.size_hint(), self.top.is_some() as usize)
     }
 
+    fn count(self) -> usize {
+        self.iter.count() + (self.top.is_some() as usize)
+    }
+
+    fn last(self) -> Option<Self::Item> {
+        self.iter.last().or(self.top)
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        if self.top.is_some() {
+            self.top = None;
+            self.iter.nth(n - 1)
```

Comment:
must check for n == 0 here. (Top could be the thing we need to return)

Reply:
Whoops!

Reply:
A quickcheck test would be a good way to test it